### PR TITLE
Viewer

### PR DESCRIPTION
**Pull request by stefanozampini**

### CODE REVIEW

File: src/binding/petsc4py/src/petsc4py/PETSc/Viewer.pyx
Comment (nocollier) [inline src/binding/petsc4py/src/petsc4py/PETSc/Viewer.pyx:92]:
```suggestion
Viewers can be called as functions where the argument specified is the PETSc object to be viewed. See the example below.

Examples
--------------
>>> from petsc4py import PETSc
>>> u = PETSc.Vec().createWithArray([1,2])
>>> v = PETSc.Viewer()
>>> v(u)
Vec Object: 1 MPI process
  type: seq
1.
2.
```

File: src/binding/petsc4py/src/petsc4py/PETSc/Viewer.pyx
Comment (nocollier) [inline src/binding/petsc4py/src/petsc4py/PETSc/Viewer.pyx:115]: This is fine but won't get rendered anywhere. For other objects like this, we added a note in the class docstring with an example. I have added this as a suggestion.

File: src/binding/petsc4py/src/petsc4py/PETSc/Viewer.pyx
Comment (nocollier) [inline src/binding/petsc4py/src/petsc4py/PETSc/Viewer.pyx:130]:
```suggestion
          If none of the above applies, it assumes ``obj`` is an instance of `Object`
```

File: src/binding/petsc4py/src/petsc4py/PETSc/Viewer.pyx
Comment (nocollier) [inline src/binding/petsc4py/src/petsc4py/PETSc/Viewer.pyx:467]:
```suggestion
    def pushFormat(self, format: Format) -> None:
```

File: src/binding/petsc4py/src/petsc4py/PETSc/Viewer.pyx
Comment (nocollier) [inline src/binding/petsc4py/src/petsc4py/PETSc/Viewer.pyx:553]:
```suggestion
        """Return the standard error viewer associated with the communicator.
```

File: src/binding/petsc4py/src/petsc4py/PETSc/Viewer.pyx
Comment (nocollier) [inline src/binding/petsc4py/src/petsc4py/PETSc/Viewer.pyx:592]:
```suggestion
        """Return the default `Type.BINARY` viewer associated with the communicator.
```

File: src/binding/petsc4py/src/petsc4py/PETSc/Viewer.pyx
Comment (nocollier) [inline src/binding/petsc4py/src/petsc4py/PETSc/Viewer.pyx:610]:
```suggestion
        """Return the default `Type.DRAW` viewer associated with the communicator.
```

File: src/binding/petsc4py/src/petsc4py/PETSc/Viewer.pyx
Comment (nocollier) [inline src/binding/petsc4py/src/petsc4py/PETSc/Viewer.pyx:642]:
```suggestion
        """Return the ASCII tab level.
```

File: src/binding/petsc4py/src/petsc4py/PETSc/Viewer.pyx
Comment (nocollier) [inline src/binding/petsc4py/src/petsc4py/PETSc/Viewer.pyx:656]:
```suggestion
        """Increment the ASCII tab level.
```

File: src/binding/petsc4py/src/petsc4py/PETSc/Viewer.pyx
Comment (nocollier) [inline src/binding/petsc4py/src/petsc4py/PETSc/Viewer.pyx:669]:
```suggestion
        """Decrement the ASCII tab level.
```

File: src/binding/petsc4py/src/petsc4py/PETSc/Viewer.pyx
Comment (nocollier) [inline src/binding/petsc4py/src/petsc4py/PETSc/Viewer.pyx:708]:
```suggestion
        """Push an additional tab level.
```

File: src/binding/petsc4py/src/petsc4py/PETSc/Viewer.pyx
Comment (nocollier) [inline src/binding/petsc4py/src/petsc4py/PETSc/Viewer.pyx:720]:
```suggestion
        """Pop an additional tab level pushed via `pushASCIITab`.
```

File: src/binding/petsc4py/src/petsc4py/PETSc/Viewer.pyx
Comment (nocollier) [inline src/binding/petsc4py/src/petsc4py/PETSc/Viewer.pyx:745]:
```suggestion
        """Print a message.
```

File: src/binding/petsc4py/src/petsc4py/PETSc/Viewer.pyx
Comment (nocollier) [inline src/binding/petsc4py/src/petsc4py/PETSc/Viewer.pyx:758]:
```suggestion
    def printfASCIISynchronized(self, msg: str) -> None:
```

File: src/binding/petsc4py/src/petsc4py/PETSc/Viewer.pyx
Comment (nocollier) [inline src/binding/petsc4py/src/petsc4py/PETSc/Viewer.pyx:759]:
```suggestion
        """Print a synchronized message.
```

File: src/binding/petsc4py/src/petsc4py/PETSc/Viewer.pyx
Comment (nocollier) [inline src/binding/petsc4py/src/petsc4py/PETSc/Viewer.pyx:786]:
```suggestion
    def setFileMode(self, mode: FileMode | str) -> None:
```

File: src/binding/petsc4py/src/petsc4py/PETSc/Viewer.pyx
Comment (nocollier) [inline src/binding/petsc4py/src/petsc4py/PETSc/Viewer.pyx:812]:
```suggestion
    def setFileName(self, name: str) -> None:
```

**Review by nocollier — COMMENTED** [review]
Lisandro can review too if you want. These are mostly minor comments with one larger suggestion adding a note about `call` to the Viewer docstring.

**Comment by stefanozampini** [comment]
thanks @nocollier . Implemented your changes. This can be merged if @dalcinl approves

File: src/binding/petsc4py/src/petsc4py/PETSc/Viewer.pyx
Comment (dalcinl) [inline src/binding/petsc4py/src/petsc4py/PETSc/Viewer.pyx:112]: If things are working, then you probably do not need this two new assignments in here.

Comment (stefanozampini) [reply src/binding/petsc4py/src/petsc4py/PETSc/Viewer.pyx]: The problem (at least what I have understood) is that with the new documentation, we will have Screen Shot 2023-04-03 at 9 08 41 PM. But then in code, you can access those enumerations via `Viewer.Mode.blah`, instead of `Viewer.FileMode.blah` as the documentation would suggest. Or am I wrong?

Comment (dalcinl) [reply src/binding/petsc4py/src/petsc4py/PETSc/Viewer.pyx]: But if you have `FileMode` at the top-level namespace, then writing `mode: FileMode | str` in function signature should resolve to the global `FileMode` outside of the `Viewer` class.

Let me clone your branch and take a final decision on this.

**Review by dalcinl — APPROVED** [review]

**Pull request merged**
